.github/workflows/ci.yml

```diff
@@ -34,6 +34,75 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      - name: Setting build env
+        run: |
+          echo "HOST_TARGET=${{ matrix.host_target }}" >> $GITHUB_ENV
+        shell: bash
+
+      # We install gnu-tar because BSD tar is buggy on Github's macos machines.
+      # See <https://github.com/actions/cache/issues/403>.
+      - name: Install GNU tar
+        if: runner.os == 'macOS'
+        run: |
+          brew install gnu-tar
+          echo "/usr/local/opt/gnu-tar/libexec/gnubin" >> $GITHUB_PATH
+
+      # Cache the global cargo directory, but NOT the local `target` directory which
+      # we cannot reuse anyway when the nightly changes (and it grows quite large
+      # over time).
+      - name: Add cache for cargo
+        uses: actions/cache@v2
+        with:
+          path: |
+            # Taken from <https://doc.rust-lang.org/nightly/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci>.
+            ~/.cargo/bin
+            ~/.cargo/registry/index
+            ~/.cargo/registry/cache
+            ~/.cargo/git/db
+            # contains package information of crates installed via `cargo install`.
+            ~/.cargo/.crates.toml
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: ${{ runner.os }}-cargo
+
+      # We use the "stable" toolchain for better caching, it is just used to build `rustup-toolchain-install-master`.
+      # But we also need to take into account that the build cache might have a different, outdated default.
```

Comment:
This is not true any more now that we do not cache `.rustup`. So this comment and the `rustup default stable` can go away I think.

And then given all that, I think we should just build xargo and rustup-toolchain-install-master with the "master toolchain". The "better caching" this refers to does not apply here anyway I think due to the different architecture.

And this would simplify things as we end up with just a single toolchain.

Reply:
I prefer building xargo and rustup-toolchain-install-master with stable if we can: to bootstrap.
We have to have rustup-toolchain-install-master to install master toolchain, don't we ?

Reply:
> We have to have rustup-toolchain-install-master to install master toolchain, don't we ?

:rofl: you are right, I proposed a cyclic dependency.^^ So disregard that part.

Do we even have to install anything though, does it not come with a stable toolchain? Who's installing `rustup` here and with which toolchain?

Reply:
> Do we even have to install anything though, does it not come with a stable toolchain? Who's installing rustup here and with which toolchain?

Yeah, [GHA does preinstalls rustup and stable toolchain, along with some CLIs](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#rust-tools), but sometimes the stable toolchain is outdated a few weeks after new Rust release. So we might want to keep the command `rustup toolchain install stable`
to update when necessary.

Reply:
Shouldn't we use `rustup update` then?

But anyway the most elegant option seems to be to use https://github.com/actions-rs/toolchain/. Thenw e can also easily set the profile to `minimal`, and we can add the rustc hash to the cache key. (We cannot cache `,rustup` as https://github.com/actions-rs/toolchain/ runs before the cache even gets installed, but this will correctly re-build xargo when there is a new stable release.)

Reply:
Or alternatively we could just use the stable that is shipped in the GHA image -- that is likely how they intend to avoid hammering the rust-lang servers unnecessarily. We don't need to always have the latest stable, it is okay if this lags by a few weeks.

Reply:
Thanks, using `rustup update` instead of `rustup toolchain install` to make the intend clearer.

Reply:
I am confused, that was not one of the two options I described above.^^ I think we should either use whatever GHA ships, or we should use https://github.com/actions-rs/toolchain/ to get the latest stable (and then incorporate the Rust version into the cache key). Why hand-roll things?

Reply:
I don't think updating the toolchain will rebuild installed latest version of CLI.
So incorporating rustc_hash for cache key may not be needed.

Removed the rustup update part.

Reply:
Yeah, adding this to the cache key would just ensure that we rebuild xargo and rustup-toolchain-install-master (I can see why you started calling it RTIM^^) when Rust is updated.

But the current solution also works, thanks. :)